lightning-net-tokio/src/lib.rs

```diff
@@ -1,279 +1,598 @@
-extern crate bytes;
-extern crate tokio;
-extern crate tokio_codec;
-extern crate futures;
-extern crate lightning;
-extern crate secp256k1;
+//! A socket handling library for those running in Tokio environments who wish to use
+//! rust-lightning with native TcpStreams.
+//!
+//! Designed to be as simple as possible, the high-level usage is almost as simple as "hand over a
+//! TcpStream and a reference to a PeerManager and the rest is handled", except for the
+//! [Event](../lightning/util/events/enum.Event.html) handlng mechanism, see below.
+//!
+//! The PeerHandler, due to the fire-and-forget nature of this logic, must be an Arc, and must use
+//! the SocketDescriptor provided here as the PeerHandler's SocketDescriptor.
+//!
+//! Three methods are exposed to register a new connection for handling in tokio::spawn calls, see
+//! their individual docs for more. All three take a
+//! [mpsc::Sender<()>](../tokio/sync/mpsc/struct.Sender.html) which is sent into every time
+//! something occurs which may result in lightning [Events](../lightning/util/events/enum.Event.html).
+//! The call site should, thus, look something like this:
+//! ```
+//! use tokio::sync::mpsc;
+//! use tokio::net::TcpStream;
+//! use secp256k1::key::PublicKey;
+//! use lightning::util::events::EventsProvider;
+//! use std::net::SocketAddr;
+//! use std::sync::Arc;
+//!
+//! // Define concrete types for our high-level objects:
+//! type TxBroadcaster = Arc<dyn lightning::chain::chaininterface::BroadcasterInterface>;
+//! type ChannelMonitor = Arc<lightning::ln::channelmonitor::SimpleManyChannelMonitor<lightning::chain::transaction::OutPoint, lightning::chain::keysinterface::InMemoryChannelKeys, TxBroadcaster>>;
+//! type ChannelManager = Arc<lightning::ln::channelmanager::ChannelManager<lightning::chain::keysinterface::InMemoryChannelKeys, ChannelMonitor, TxBroadcaster>>;
+//! type PeerManager = Arc<lightning::ln::peer_handler::PeerManager<lightning_net_tokio::SocketDescriptor, ChannelManager>>;
+//!
+//! // Connect to node with pubkey their_node_id at addr:
+//! async fn connect_to_node(peer_manager: PeerManager, channel_monitor: ChannelMonitor, channel_manager: ChannelManager, their_node_id: PublicKey, addr: SocketAddr) {
+//!     let (sender, mut receiver) = mpsc::channel(2);
+//!     lightning_net_tokio::connect_outbound(peer_manager, sender, their_node_id, addr).await;
+//!     loop {
+//!         receiver.recv().await;
+//!         for _event in channel_manager.get_and_clear_pending_events().drain(..) {
+//!             // Handle the event!
+//!         }
+//!         for _event in channel_monitor.get_and_clear_pending_events().drain(..) {
+//!             // Handle the event!
+//!         }
```

Comment:
The drain(..) means you have ownership of the events in the loop instead of a reference to. I guess its not strictly necessary but it does make things easier to work with often.